Allabana/tests

piraka9011 opened the pull request:
Tests for the sysroot creator. 
Note that #1 should be merged first before this PR and then changes from there remerged here.

Commit:
Added spinner to docker build image

Commit:
Spinners, Requirements, Docs

Commit:
Install script and doc updates

Commit:
Update to output Docker stdout to tool output

Commit:
Modify argparse code to check for argument validity better

Commit:
Update for CR:
- Updated mixins to reflect toolchain file
- Address nits
- Fix docs
- Updated requirements
- Switch print to logging
TODO: Update default Dockerfile args.

Commit:
- Automatically setup build environment at the end of script
- Add progress bar for main pipeline

Commit:
Temp. update for some tests.
- Change directory format to conform with ROS standard
- Create python package directroy structure
- Initial test

Commit:
Temp. update for some tests.
- Change directory format to conform with ROS standard
- Create python package directroy structure
- Initial test

Commit:
Sysroot creator tests - untested

Commit:
Sysroot creator tests - untested

Commit:
Update README

Commit:
Refactor tests

Commit:
Refactor tests

Commit:
Update Dockerfile defaults

Commit:
Change logger to info

piraka9011 committed:
Merge branch 'cc-sysroot-creator' into allabana/cc-sysroot-creator

piraka9011 committed:
README hotfix

Commit:
Update README.md

thomas-moulard commented on tests/test_sysroot_creator.py:
```diff
@@ -0,0 +1,191 @@
+import docker
+
+
+__author__ = 'Anas Abou Allaban'
```
Let's drop those metadata as they don't really help (+we maintain code as a team and don't want to expose personal ownership)

thomas-moulard commented on tests/test_sysroot_creator.py:
```diff
@@ -0,0 +1,191 @@
+        network_mode='host', nocache=False)
+
+
+@pytest.mark.run(order=1)
```
Any reason to order those tests?

thomas-moulard commented on README.md:
````diff
@@ -26,6 +26,26 @@ newgrp docker # this reloads the group permissions in the current shell, unneces
+# Copy ROS Sources
+cp -r ~/ros2_ws/src .
+```
+In the end your Directory should be as follows:
````
```suggestion
In the end your directory should be as follows:
```

thomas-moulard commented on INSTALL_MAC.md:
```diff
@@ -0,0 +1,61 @@
+Additional base images can be created using the script `colcon_cc_build/colcon_cc_build/verb/sysroot/publish_to_ECR.sh`.
+Some base images already available to use with `--sysroot-base-image`:
+
+- `juanrh/ros2:armhf_bionic_crystal_fastrtps_prebuilt-crystal` for building ROS 2 packages starting from a prebuilt binary for ROS 2 crystal for armhf.
```
Let's add a FIXME for all those mentions to juanrh to avoid forgeting about it.

piraka9011 replied:
Fixed in PR #1

thomas-moulard commented on mixins/cross-compile.mixin:
```diff
@@ -0,0 +1,50 @@
+    "build": {
+        "armhf-generic_linux": {
+            "cmake-args": [
+                "-DCMAKE_SYSTEM_NAME=Linux",
```
There is some overhead between the mixin and armhf-linux.cmake we should resolve.

piraka9011 replied:
Fixed in PR #1

thomas-moulard commented on mixins/lint.py:
```diff
@@ -0,0 +1,53 @@
+"""
+Lint the index.yaml file as well as all files ending with .mixin.
+Taken from https://github.com/colcon/colcon-mixin-repository.
+Modified for the AWS mixins repository.
```
Please make sure that the licenses are compatible.

piraka9011 replied:
Pre-approved per [Amazon Opensource Preapproved List](https://w.amazon.com/index.php/Open_Source/Distributions/Process/License_Instructions_for_Distribution)

thomas-moulard commented on sysroot_creator/scripts/create_cc_sysroot.py:
```diff
@@ -44,6 +48,8 @@
 SYSROOT_DIR_NAME: str = 'sysroot'
 DOCKER_WS_NAME: str = 'Dockerfile_workspace'
 DOCKER_CLIENT = docker.from_env()
+logging.basicConfig(level=logging.INFO)
```
Let's avoid having code executed at the top level - this can be moved to the main function (block).

thomas-moulard commented on sysroot_creator/scripts/create_cc_sysroot.py:
```diff
@@ -129,52 +130,69 @@ def build_workspace_sysroot_image(
+            network_mode=docker_args.network_mode,
+            decode=True)
+        # Progress bar with postfix
+        with tqdm(log_generator) as t:
```
Coudl we split that to another function? seems like a lot of things are happening there.

thomas-moulard commented on sysroot_creator/scripts/create_cc_sysroot.py:
```diff
@@ -129,52 +130,69 @@ def build_workspace_sysroot_image(
+        with tqdm(log_generator) as t:
+            for chunk in t:
+                try:
+                    # Remove carriage returns and new lines ('\r\n')
```
Seems like this could be a separate utility function.
Do we have unit tests for those? Making those top-level utility functions would help us making sure all this string manipulation code works well.

thomas-moulard commented on sysroot_creator/scripts/create_cc_sysroot.py:
```diff
@@ -129,52 +130,69 @@ def build_workspace_sysroot_image(
+                        continue
+                    t.set_postfix_str(line)
+                except (KeyError, AttributeError) as e:
+                    logger.exception("Error reading the Dockerfile image. The following "
```
nit: logger.exception will display the exception/backtrace automatically so you don't need to pass `e`

thomas-moulard commented on sysroot_creator/scripts/create_cc_sysroot.py:
```diff
@@ -129,52 +130,69 @@ def build_workspace_sysroot_image(
+        logger.exception('Error building sysroot image. The following exception was '
+                         'caught:\n{}'.format(be.msg))
+        logger.exception('Build log:')
+        for stream_obj in be.build_log:
```
sounds like this should be a separate function

thomas-moulard commented on sysroot_creator/scripts/create_cc_sysroot.py:
```diff
@@ -219,56 +237,87 @@ def write_cc_system_setup_script(cc_root_dir: Path, platform: Platform):
     return parser
 
 
+def update_pbar(pbar: tqdm, step: int) -> (tqdm, int):
```
pbar -> progress_bar
Avoiding abbreviations let other users (or you in 6+ months) understand the code more easily.

thomas-moulard commented on sysroot_creator/scripts/create_cc_sysroot.py:
```diff
@@ -219,56 +237,87 @@ def write_cc_system_setup_script(cc_root_dir: Path, platform: Platform):
 
 
+def update_pbar(pbar: tqdm, step: int) -> (tqdm, int):
+    step += 1
```
This is weird no? Can't you get the current step from the object and increment it? I think this behavior warrants some documentation. From an outside perspective, from the function signature I would expect this to set the progress to "step", not "step+1".

thomas-moulard commented on tests/test_sysroot_creator.py:
```diff
@@ -0,0 +1,191 @@
+#!/usr/bin/env python3
```
Unit tests should generally not touch the file system. Is this the case here?
If yes, consider using pyfakefs (https://github.com/jmcgeheeiv/pyfakefs).

thomas-moulard reviewed (commented):
This generally LGTM! A few nits. I'll let Prajakta giving the final green light.

piraka9011 committed:
Comments & hotfix
- Added some comments
- Fix for capturing end of dockerfile output stream.

piraka9011 committed:
Cleanup with bug fixes
- Updated Readme conflicts
- Update dockerfile to address cahcing error with rosdep
- Removed tqdm progress bar and cleaned up output stream for dockerfile API

piraka9011 committed:
Merge remote-tracking branch 'origin/allabana/cc-sysroot-creator' into allabana/tests

# Conflicts:
#	sysroot_creator/scripts/create_cc_sysroot.py

prajakta-gokhale commented:
Could you please rebase this PR on top of `cc-sysroot-creator` so only files that need to be merged are in the review diff?

Commit:
Changes to requirements and gitignore

Signed-off-by: Prajakta Gokhale <prajaktg@amazon.com>

Commit:
Changes to colcon mixins

* Add aarch64 mixin
* Rename mixins to <arch>-<platform>
* Add more build flags

Signed-off-by: Prajakta Gokhale <prajaktg@amazon.com>

Commit:
Reflect changes in README.md

Signed-off-by: Prajakta Gokhale <prajaktg@amazon.com>

Commit:
Merge remote-tracking branch 'origin/allabana/cc-sysroot-creator' into cc-sysroot-creator

piraka9011 committed:
PR Updates
- Add argumnent assertions
- Cleanup tests to be only unit tests
- Add scripts and directory cleanup

piraka9011 committed:
PR Updates
- Add argumnent assertions
- Cleanup tests to be only unit tests
- Add scripts and directory cleanup

piraka9011 committed:
Merge branch 'allabana/tests' of github.com:aws-ros-dev/cross_compile into allabana/tests

prajakta-gokhale commented on sysroot_creator/scripts/create_cc_sysroot.py:
```diff
@@ -101,6 +99,8 @@ def __str__(self):
 
 
 def setup_cc_root_dir(platform: Platform) -> Path:
+    if not isinstance(platform, Platform):
```
Curious why this check is added - when do we expect this to not be of type `Platform`?

Also with type hints, do we need to check it at all?

piraka9011 replied:
No, but you can still pass bad inputs, so I added this to make the tests pass when bad data is sent.
We aren't expecting the type to change though.

prajakta-gokhale commented on sysroot_creator/scripts/create_cc_sysroot.py:
```diff
@@ -234,12 +243,13 @@ def write_cc_system_setup_script(cc_root_dir: Path, platform: Platform):
-    logger.debug("Executing 'source {}'".format(build_setup_path))
-    subprocess.run(['source', build_setup_path])
+    logger.info("Executing 'bash {}'".format(sys_setup_path))
+    subprocess.run(['bash', str(sys_setup_path)])
```
Not entirely sure, but I think this will require sudo/password prompt?

piraka9011 replied:
Hasn't requested it for me... (Mac 10.14/Ubuntu 18.04 EC2)

prajakta-gokhale commented on tests/test_sysroot_creator.py:
```diff
@@ -0,0 +1,140 @@
+"""Tests for the `create_cc_sysroot.py` script.
+
+Test modules are executed in sequence and depend on one another to emulate the
```
Nit: Please update this comment.

prajakta-gokhale reviewed (commented):
Looks good to me overall, thanks for rebasing. A few minor comments below.

Also, have you run linting on this?

piraka9011 committed:
PR changes and cleanup
